test/src/unit-vfs.cc

```diff
@@ -351,3 +352,86 @@ TEST_CASE("VFS: test ls_with_sizes", "[vfs][ls-with-sizes]") {
   // Clean up
   REQUIRE(vfs_ls.remove_dir(URI(path)).ok());
 }
+
+// Currently only S3 is supported for VFS::ls_cb.
+using TestBackends = std::tuple<S3Test>;
+TEMPLATE_LIST_TEST_CASE(
+    "VFS: Test internal ls_cb recursion argument",
+    "[vfs][ls_cb]",
+    TestBackends) {
+  TestType fs({10, 50});
+  if (!fs.is_supported()) {
+    return;
+  }
+
+  tiledb::sm::LsCallbackCAPI cb =
+      [](const char* path, size_t path_len, uint64_t size, void* data) {
```

Comment:
The selection predicates for our own tests should not participate in the C API. IOW there should be no `data` argument.